queue/queueservice.js

```diff
@@ -303,15 +315,140 @@ class QueueService {
     ].join('-');
 
     // Return and cache promise that we created this queue
-    return this.queues[id] = this._createQueue(name).catch(err => {
+    return this.queues[id] = Promise.all([
+      this._ensureQueueAndMetadata(name, provisionerId, workerType),
+      this._ensureQueueAndMetadata(legacyName, provisionerId, workerType)
+    ]).catch(err => {
+      debug("[alert-operator] Failed to ensure azure queue: %s, as JSON: %j",
+            err, err, err.stack);
+
       // Don't cache negative results
       this.queues[id] = undefined;
       throw err;
-    }).then(() => {
-      return name;
-    });
+    }).then(() => name);
+  }
+
+  /**
+   * Ensure that queue with given name exists and has correct meta-data,
+   * in particular regarding meta-data for expiration.
+   *
+   * We maintain following meta-data keys:
+   *  * `provisioner_id`,
+   *  * `worker_type`, and
+   *  * `last_used` (tolerating 23 hours difference, updated every 25 hours)
+   *
+   * We delete queues if they have `last_used` > 10 days ago, this happens in
+   * a periodic test tasks.
+   */
+  async _ensureQueueAndMetadata(queue, provisionerId, workerType) {
+    // Fetch meta-data from queue, checking if it exists
+    try {
+      let {metadata} = await this.client.getMetadata(queue);
+
+      // Check if meta-data is up-to-date
+      var lastUsed = new Date(metadata.last_used);
+      if (metadata.provisioner_id === provisionerId
+          && metadata.worker_type === workerType
+          && isFinite(lastUsed)
+          && lastUsed.getTime() > Date.now() - 23 * 60 * 60 * 1000
+      ) {
+        return; // We're done as meta-data is present
```

Comment:
Is there something elsewhere that sets last used?  If we are ensuring the queue exists, does that mean we're intending to use it so last used is now?